### ISSUE TITLE
Bad sector remapping in USB MSC mode

### ISSUE THREAD

**Issue by ChrisPVille** [ChrisPVille/RL02#3]
Modern OSes get unhappy when they have to deal with IO errors from disks.  The RL02 disk packs pretty much all had bad sectors somewhere.  These can be identified and stored in a factory provided table on the last track of the disk, but it was up to each OS to decide how to handle them.  Some skip sectors while others remap them.

In USB Mass Storage mode, save a track of the RL02 for sector remapping and use the DEC written bad sector table.  With these, modern filesystems can be written to the disk around the inevitable bad sectors, although this will break SIMH and block level disk backup support for vintage disks.

**Comment by smaresca** [comment]
@ChrisPVille I'm curious: have you revisited this item by any chance?

**Comment by ChrisPVille** [comment]
Nope. I haven't worked on the hard drive project in a while.  It was becoming a major headache to constantly repair the age-related problems that kept popping up in the drive as I was trying to build this.  I did pick up a few more RL-02s in the interim, and I will get back to working on this... eventually.

As far as the bad sector map, I don't think it needs to be in the DEC format.  I suspect any drive being used in MSC mode is not ever going to be fully compatible with the PDP-11 emulators.  Because the emulated software (RT-11, etc.) handles bad sectors internally, it would be an ugly process of decoding the DEC bad sector format on the disk, managing that from the microcontroller, communicating that through a side-channel to a modified SimH which would feed that through to the guest OS where any failing writes to bad sectors would get communicated back to the uC for inclusion on that table.

It makes more sense to have MSC mode (LBA) available for modern filesystems which can handle their own bad blocks and provide a special SimH driver exposing the underlying Cylinder Head Sector format a virtual RL11 controller would have.  The microcontroller never has to worry about marking or relocating bad blocks as higher level pieces of software will do that.
